addon/webextension/background/auth.js

```diff
@@ -0,0 +1,208 @@
+    return registrationInfo && registrationInfo.deviceId;
+  };
+
+  function generateRegistrationInfo() {
+    let info = {
+      deviceId: "anon" + makeUuid() + "",
+      secret: makeUuid()+"",
+      registered: false
+    };
+    return info;
+  }
+
+  function register() {
+    return new Promise((resolve, reject) => {
+      let registerUrl = main.getBackend() + "/api/register";
+      let req = new XMLHttpRequest();
+      req.open("POST", registerUrl);
+      req.setRequestHeader("content-type", "application/x-www-form-urlencoded");
+      req.onload = catcher.watchFunction(() => {
+        if (req.status == 200) {
+          console.info("Registered login");
+          initialized = true;
+          saveAuthInfo(JSON.parse(req.responseText));
+          resolve(true);
+          analytics.sendEvent("registered");
+        } else {
+          console.warn("Error in response:", req.responseText);
+          reject(new Error("Bad response: " + req.status));
+        }
+      });
+      req.send(uriEncode({
+        deviceId: registrationInfo.deviceId,
+        secret: registrationInfo.secret,
+        deviceInfo: JSON.stringify(deviceInfo())
+      }));
+    });
+  }
+
+  function login(options) {
+    let { ownershipCheck, noRegister } = options || {};
+    return new Promise((resolve, reject) => {
+      let loginUrl = main.getBackend() + "/api/login";
+      let req = new XMLHttpRequest();
+      req.open("POST", loginUrl);
+      req.onload = catcher.watchFunction(() => {
+        if (req.status == 404) {
+          if (noRegister) {
+            resolve(false);
+          } else {
+            resolve(register());
+          }
+        } else if (req.status >= 300) {
+          console.warn("Error in response:", req.responseText);
+          reject(new Error("Could not log in: " + req.status));
+        } else if (req.status === 0) {
+          let error = new Error("Could not log in, server unavailable");
+          analytics.sendEvent("login-failed");
+          reject(error);
+        } else {
+          initialized = true;
+          let jsonResponse = JSON.parse(req.responseText);
+          console.info("Screenshots logged in");
+          analytics.sendEvent("login");
+          saveAuthInfo(jsonResponse);
+          if (ownershipCheck) {
+            resolve({isOwner: jsonResponse.isOwner});
+          } else {
+            resolve(true);
+          }
+        }
+      });
+      req.onerror = catcher.watchFunction(() => {
+        let exc = new Error("Connection failed");
+        exc.url = loginUrl;
+        exc.popupMessage = "CONNECTION_ERROR";
+        reject(exc);
+      });
+      req.setRequestHeader("content-type", "application/x-www-form-urlencoded");
+      req.send(uriEncode({
+        deviceId: registrationInfo.deviceId,
+        secret: registrationInfo.secret,
+        deviceInfo: JSON.stringify(deviceInfo()),
+        ownershipCheck
+      }));
+    });
+  }
+
+  function saveAuthInfo(responseJson) {
```

Comment:
So... here's how I think about what we're doing here:

1. The real canonical location for this information is in memory in the background page.  We initialize the values eagerly (at the module level), and update them in-place while also recording updates in storage.  
2. We're relying on a simple dumb store (i.e., dropping in `storage.sync` would break some assumptions).  That is, nothing ever changes behind our back.
3. `registrationInfo` is a compound object, so there's some danger we're overwriting something that we meant to keep – except for the fact that its canonical location is in memory.
4. `abTests` and `registrationInfo` aren't really related.  Really `abTests` is canonically stored on the server, but we can't fetch it eagerly, so the storage here is more like a cache.
5. The real concern I see is that we overwrite registrationInfo when it has already been used.  I think we're pretty safe here, assuming that `storage.local` is consistent and reliable.  We only set `registrationInfo` at module-load time, so there's no real opportunity to overwrite it so long as `storage.local` is reliable.  *Except...*
6. We also set the info in `setDeviceInfoFromOldAddon()`.  Or rather we do it wrong, ugh.  Filed as #2588 –
 anyway, that also happens at module load time based on communication with bootstrap.js